### PR TITLE
CorfuStoreBrowser: Allow addRecord on null metadata tables

### PR DESCRIPTION
**Pull request by hisundar**
## Overview

Description:

Why should this be merged: 
If the table does not have a metadata section browser's addRecord fails with a NullPointerException

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: test/src/test/java/org/corfudb/integration/CorfuStoreBrowserEditorIT.java [hunk 1]
```diff
@@ -861,6 +870,10 @@ public void addRecordTest() throws Exception {
             TABLE_NAME, newKeyString, newValString, newMetadataString);
         Assert.assertNotNull(addedRecord);
 
+        CorfuDynamicRecord addedRecord2 = browser.addRecord(NAMESPACE,
+                table2Name, newKeyString, newValString, newMetadataString);
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/integration/CorfuStoreBrowserEditorIT.java:874]: need to pass null metadata here, right?

**Review by pankti-m — COMMENTED** [review]
Some tables have metadata type as 'Message'.  What should be the input to addRecord() in that case?  Can we test that case also?

**Review by shama358 — COMMENTED** [review]
the listStreamTag and getTagToTableNamesMap look into the metadata msg. Should we have a check there as well?

**Comment by hisundar** [comment]
> Some tables have metadata type as 'Message'. What should be the input to addRecord() in that case? Can we test that case also?

For Message - an empty `'{}'` seems to work, but I will add a test case for that too.

**Comment by hisundar** [comment]
> the listStreamTag and getTagToTableNamesMap look into the metadata msg. Should we have a check there as well?

those are looking at the Metadata from the TableRegistry entry - which is not going to be null right? The bug only arises when someone has a null in their own table's metadata section.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java [hunk 2]
```diff
@@ -408,8 +408,9 @@ public CorfuDynamicRecord addRecord(String namespace, String tableName,
             dynamicProtobufSerializer.createDynamicMessageFromJson(defaultValueAny,
                 newValue);
         DynamicMessage newMetadataMsg =
+                !defaultMetadataAny.getTypeUrl().isEmpty() ?
```
[the remark this answers is not in the document]
Comment (hisundar) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java:411]: I ran this through the debugger and what I found is the AnyType does not set the typeUrl to null instead it sets it to an empty string. that's why I also test here for empty string whenever the metadata is null.

File: test/src/test/java/org/corfudb/integration/CorfuStoreBrowserEditorIT.java (hunk 1 above)
Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/CorfuStoreBrowserEditorIT.java:874]: both null and non-null works here since the metadata of the table is taken into account. But for clarity I can change this to be null as well.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java:411]: but can defaultMetadataAny be null?  Can you please add a comment that defaultMetadataAny cannot be null?  It is a bit difficult to determine from the code..  A comment that typeUrl is an empty string in this case will also be helpful.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java [hunk 3]
```diff
@@ -408,8 +408,9 @@ public CorfuDynamicRecord addRecord(String namespace, String tableName,
             dynamicProtobufSerializer.createDynamicMessageFromJson(defaultValueAny,
                 newValue);
         DynamicMessage newMetadataMsg =
```
Comment (pankti-m) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java:410]: instead of adding this validation, can we directly set newMetadataMsg to null if the input argument is null?

This did not occur to me yesterday but if you agree with the approach, it will make the change simpler.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuDynamicRecord.java [hunk 4]
```diff
@@ -73,8 +78,10 @@ public boolean equals(Object obj) {
         boolean payloadMatch = payloadTypeUrl.equals(corfuDynamicRecord.payloadTypeUrl)
                 && payload.equals(corfuDynamicRecord.payload);
 
-        // Compare the metadata.
-        boolean metadataMatch = metadata.equals(corfuDynamicRecord.metadata);
+        // Compare the metadata only if both are not null.
+        boolean metadataMatch = (metadata != null && corfuDynamicRecord.metadata != null) ?
```
Comment (pankti-m) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuDynamicRecord.java:82]: minor:
metadataMatch = metadata == corfuDynamicRecord.metadata || metadata.equals(..) will make this check simpler

File: test/src/test/java/org/corfudb/integration/CorfuStoreBrowserEditorIT.java [hunk 5]
```diff
@@ -869,9 +883,13 @@ public void addRecordTest() throws Exception {
         String metadataTypeUrl = Any.pack(newMetadataUuid).getTypeUrl();
         CorfuDynamicRecord expectedRecord = new CorfuDynamicRecord(valTypeUrl,
             dynamicValMessage, metadataTypeUrl, dynamicMetadataMessage);
-
         Assert.assertEquals(expectedRecord, addedRecord);
         Assert.assertEquals(2, browser.printTable(NAMESPACE, TABLE_NAME));
+
+        CorfuDynamicRecord expectedRecord2 = new CorfuDynamicRecord(valTypeUrl,
+                dynamicValMessage, null, null);
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/integration/CorfuStoreBrowserEditorIT.java:890]: if metadata is null, the typeUrl is an empty string instead of a null.. I suggest extracting the typeUrl like L883 so that the test passes values like the actual code.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java (hunk 3 above)
Comment (hisundar) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java:410]: Doesn't that change the semantics of what this `addRecord()` does - currently if the table had metadata configured then we will expect a metadata to be passed in. But now if we just validate it on the incoming metadata (being null or not) it means the users can pass null metadata in a table which had non-null metadata configured right?

**Comment by hisundar** [comment]
>but can defaultMetadataAny be null? Can you please add a comment that defaultMetadataAny cannot be null? It is a bit >difficult to determine from the code.. A comment that typeUrl is an empty string in this case will also be helpful.
no, even when the table does not have any metadata configured (user passes in null) this `defaultMetadataAny` is not null here, because it is reading from the registry table's section which has other non-null fields like key, so it returns a default instance of the Any type instead of a null for metadata section.
I will add a comment explaining this. thanks

File: test/src/test/java/org/corfudb/integration/CorfuStoreBrowserEditorIT.java (hunk 5 above)
Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/CorfuStoreBrowserEditorIT.java:890]: Sure, I will fix the expectedRecord the same way.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuDynamicRecord.java (hunk 4 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuDynamicRecord.java:82]: done.

**Review by shama358 — APPROVED** [review]
LGTM

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java (hunk 3 above)
Comment (hisundar) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserEditor.java:420]: discussed offline - implemented the validation to fail if table had metadata configured but user attempts to pass null.

**Review by pankti-m — APPROVED** [review]
